tests/Unit/NumericalAlgorithms/RootFinding/Test_RootBracketing.cpp

```diff
@@ -0,0 +1,155 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "Framework/TestingFramework.hpp"
+
+#include <boost/none.hpp>
+#include <boost/optional.hpp>
+#include <cmath>
+#include <random>
+
+#include "DataStructures/DataVector.hpp"
+#include "Framework/TestHelpers.hpp"
+#include "NumericalAlgorithms/RootFinding/RootBracketing.hpp"
+#include "Utilities/ConstantExpressions.hpp"
+#include "Utilities/Gsl.hpp"
+
+namespace {
+boost::optional<double> f_free(double x) noexcept {
+  return (x < 1.0 or x > 2.0) ? boost::none
+                              : boost::optional<double>(2.0 - square(x));
+}
+struct F {
+  boost::optional<double> operator()(double x) const noexcept {
+    return (x < 1.0 or x > 2.0) ? boost::none
+                                : boost::optional<double>(2.0 - square(x));
+  }
+};
+
+template <typename Function>
+void test_bracketing_simple_one_function(
+    const Function& f, const std::array<double, 2>& bounds,
+    const boost::optional<double>& guess) noexcept {
+  double lower = bounds[0];
+  double upper = bounds[1];
+  double f_at_lower = std::numeric_limits<double>::signaling_NaN();
+  double f_at_upper = std::numeric_limits<double>::signaling_NaN();
+  if (guess) {
+    RootFinder::bracket_possibly_undefined_function_in_interval(
+        &lower, &upper, &f_at_lower, &f_at_upper, f, guess.get());
+  } else {
+    RootFinder::bracket_possibly_undefined_function_in_interval(
+        &lower, &upper, &f_at_lower, &f_at_upper, f);
+  }
+  CHECK(f_at_lower * f_at_upper <= 0.0);
+  CHECK(lower <= sqrt(2.0));
+  CHECK(upper >= sqrt(2.0));
+}
+
+void test_bracketing_simple_multiple_functions(
+    const std::array<double, 2>& bounds,
+    const boost::optional<double>& guess) noexcept {
+  const auto f_lambda = [](double x) noexcept->boost::optional<double> {
+    return (x < 1.0 or x > 2.0) ? boost::none
+                                : boost::optional<double>(2.0 - square(x));
+  };
+  const F f_functor{};
+
+  test_bracketing_simple_one_function(f_free, bounds, guess);
+  test_bracketing_simple_one_function(f_lambda, bounds, guess);
+  test_bracketing_simple_one_function(f_functor, bounds, guess);
+}
+
+void test_bracketing_simple() noexcept {
+  // Set up random number generator
+  MAKE_GENERATOR(gen);
+  std::uniform_real_distribution<> unit_dis(0.0, 1.0);
+
```

Comment:
Is it true that this does not test the case where the control flow reaches the bounds-contracting helper function and all of the array members are valid, but of the same sign? If not, I think that case (where the sign changes multiple times in the domain and all of the bounds and guess start with the same sign in the valid region) should be tested.